### PR TITLE
Add timeout for Cassandra Schema creation job

### LINKED ISSUE #818: Jaeger Operator not deploying Jaeger components
I recently noticed that my Jaeger Operator (version **1.14.0**) doesn't deploy my Jaeger CRD definition anymore, i.e., no Jaeger components are created (collector, query). However, the Cassandra pods get created, and the Cassandra Schema job get executed.

Having a look into the operator's logs, I can see a lot of lines looking like that:

```
time="2019-12-05T16:14:53Z" level=error msg="failed to apply the changes" error="failed to handler dependencies: timed out waiting for the condition" execution="2019-12-05 16:12:53.593407803 +0000 UTC" instance=jaeger namespace=tracing
```

Looking inside the operator's source code, I couldn't find out where this error was coming from. Would you have an idea of what is happening?

I deploy Jaeger Operator with the official Helm chart (version **2.10.2**). And here is my Jaeger CRD:

```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
  labels:
    app.kubernetes.io/name: jaeger
    helm.sh/chart: jaeger-0.1.0
    app.kubernetes.io/instance: jaeger
    app.kubernetes.io/managed-by: Tiller
spec:
  collector:
    image: jaegertracing/jaeger-collector:1.14
    replicas: 1
    resources:
      limits:
        cpu: 1
        memory: 154Mi
      requests:
        cpu: 1
        memory: 128Mi
    tolerations:
      - effect: NoSchedule
        key: CriticalAddonsOnly
        operator: Exists
  ingress:
    enabled: false
  query:
    image: jaegertracing/jaeger-query:1.14
    replicas: 1
    resources:
      limits:
        cpu: 0.1
        memory: 128Mi
      requests:
        cpu: 0.01
        memory: 64Mi
    tolerations:
      - effect: NoSchedule
        key: CriticalAddonsOnly
        operator: Exists
  storage:
    cassandraCreateSchema:
      datacenter: dc1
      mode: test
    options:
      cassandra:
        keyspace: jaeger_v1_dc1
        servers: jaeger-cassandra.tracing.svc.cluster.local
    type: cassandra
  strategy: production
```

Once created on Kubernetes, it then looks like that (`kubectl get jaeger jaeger -o yaml`):

```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  creationTimestamp: "2019-12-05T15:35:18Z"
  generation: 2
  labels:
    app.kubernetes.io/instance: jaeger
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/name: jaeger
    helm.sh/chart: jaeger-0.1.0
    jaegertracing.io/operated-by: tracing.jaeger-operator
  name: jaeger
  namespace: tracing
  resourceVersion: "269092844"
  selfLink: /apis/jaegertracing.io/v1/namespaces/tracing/jaegers/jaeger
  uid: d75eeec1-1774-11ea-a962-0a82c62d427e
spec:
  agent:
    options: {}
    resources: {}
  allInOne:
    options: {}
    resources: {}
  collector:
    image: jaegertracing/jaeger-collector:1.14
    options: {}
    replicas: 1
    resources:
      limits:
        cpu: "1"
        memory: 154Mi
      requests:
        cpu: "1"
        memory: 128Mi
    tolerations:
    - effect: NoSchedule
      key: CriticalAddonsOnly
      operator: Exists
  ingester:
    options: {}
    resources: {}
  ingress:
    enabled: false
    openshift: {}
    resources: {}
  query:
    image: jaegertracing/jaeger-query:1.14
    options: {}
    replicas: 1
    resources:
      limits:
        cpu: 100m
        memory: 128Mi
      requests:
        cpu: 10m
        memory: 64Mi
    tolerations:
    - effect: NoSchedule
      key: CriticalAddonsOnly
      operator: Exists
  resources: {}
  sampling:
    options: {}
  storage:
    cassandraCreateSchema:
      datacenter: dc1
      mode: test
    dependencies:
      resources: {}
    elasticsearch:
      storage: {}
    esIndexCleaner:
      resources: {}
    esRollover:
      resources: {}
    options:
      cassandra:
        keyspace: jaeger_v1_dc1
        servers: jaeger-cassandra.tracing.svc.cluster.local
    type: cassandra
  strategy: production
  ui:
    options: {}
status:
  version: ""
```

### PR DESCRIPTION
**Pull request by arugifa**
Cassandra pods need sometimes more than 2 minutes to bootstrap. When
this happens, the Cassandra Schema job is failing, and prevents further
deployment of Jaeger components.

By allowing to define a custome timeout for this job, we can circumvent
this problem.

Resolves #818

Signed-off-by: Alexandre Figura <arugifa@users.noreply.github.com>

### CODE REVIEW

File: pkg/storage/cassandra_dependencies.go [hunk 1]
```diff
@@ -50,8 +50,12 @@ func cassandraDeps(jaeger *v1.Jaeger) []batchv1.Job {
 		"linkerd.io/inject":       "disabled",
 	}
 
-	// TODO: should this be configurable? Would we ever think that 2 minutes is OK for this job to complete?
-	deadline := int64(120)
+	timeout := jaeger.Spec.Storage.CassandraCreateSchema.Timeout
+	if timeout == nil {
+		jaeger.Logger().Info("Timeout not specified. Using '120' for the cassandra-create-schema job.")
```
Comment (jpkrohling) [inline pkg/storage/cassandra_dependencies.go]: Perhaps debug?

File: pkg/apis/jaegertracing/v1/jaeger_types.go [hunk 2]
```diff
@@ -399,6 +399,9 @@ type JaegerCassandraCreateSchemaSpec struct {
 	// +optional
 	Mode string `json:"mode,omitempty"`
 
+	// +optional
+	Timeout *int64 `json:"timeout,omitempty"`
```
Comment (jpkrohling) [inline pkg/apis/jaegertracing/v1/jaeger_types.go]: Perhaps leave a comment that this is in line with what is being used downstream? Intuitively, I would expect a "duration", similar to what we have elsewhere here (see `ReadTTL`), but looking at the cronjob spec, I see that it accepts an integer representing the number of "seconds".

File: pkg/storage/cassandra_dependencies.go (hunk 1 above)
Comment (arugifa) [reply pkg/storage/cassandra_dependencies.go]: Done 👍

File: pkg/apis/jaegertracing/v1/jaeger_types.go (hunk 2 above)
Comment (arugifa) [reply pkg/apis/jaegertracing/v1/jaeger_types.go]: I was wondering how to handle duration parsing errors? I decided to go with using a default timeout (120s), but maybe blocking job deployment would make more sense? I couldn't find out how to test such behavior though (that's why I decided to use the default timeout).

Comment (jpkrohling) [reply pkg/apis/jaegertracing/v1/jaeger_types.go]: In the operator, we typically fall back to a sensible default + log at info/warn level whenever possible. For instance, we fallback to "in memory" storage when the storage type is set to an unknown type.

**Pull request merged**